store/unit_store.go

```diff
@@ -89,31 +93,69 @@ func (s unitStores) Refs(f ...RefFilter) ([]*graph.Ref, error) {
 	}
 
 	c_unitStores_Refs_last_numUnitsQueried = 0
-	var allRefs []*graph.Ref
+	var (
+		allRefs   []*graph.Ref
+		allRefsMu sync.Mutex
+	)
+	par := parallel.NewRun(storeFetchPar)
 	for u, us := range uss {
+		u, us := u, us
 		if us == nil {
 			continue
 		}
 
-		c_unitStores_Refs_last_numUnitsQueried++
-		setImpliedUnit(f, u)
-		refs, err := us.Refs(filtersForUnit(u, f).([]RefFilter)...)
-		if err != nil && !isStoreNotExist(err) {
-			return nil, err
-		}
-		for _, ref := range refs {
-			ref.UnitType = u.Type
-			ref.Unit = u.Name
-			if ref.DefUnitType == "" {
-				ref.DefUnitType = u.Type
+		par.Do(func() error {
+			// Copy so we can do concurrent modification in setImpliedUnit
+			fCopy := make([]RefFilter, len(f))
+			for i, filter := range f {
+				// HACK: consider adding a clone() method to filters
+				switch filter := filter.(type) {
+				case *absRefFilterFunc:
+					newFilter := *filter
+					fCopy[i] = &newFilter
+				case *byRefDefFilter:
+					newFilter := *filter
+					fCopy[i] = &newFilter
+				default:
+					if _, ok := filter.(impliedUnitSetter); ok {
+						return fmt.Errorf("cannot shallow-copy unrecognized filter type %T", filter)
+					} else {
+						fCopy[i] = filter
+					}
+				}
+			}
+			fCopy = filtersForUnit(u, fCopy).([]RefFilter)
+
+			setImpliedUnit(fCopy, u)
+
+			// Use a lock since increment isn't atomic
+			c_mu.Lock()
+			c_unitStores_Refs_last_numUnitsQueried++
+			c_mu.Unlock()
+
+			refs, err := us.Refs(fCopy...)
+			if err != nil && !isStoreNotExist(err) {
+				return err
 			}
-			if ref.DefUnit == "" {
-				ref.DefUnit = u.Name
+			for _, ref := range refs {
+				ref.UnitType = u.Type
+				ref.Unit = u.Name
+				if ref.DefUnitType == "" {
+					ref.DefUnitType = u.Type
+				}
+				if ref.DefUnit == "" {
+					ref.DefUnit = u.Name
+				}
 			}
-		}
-		allRefs = append(allRefs, refs...)
+
+			allRefsMu.Lock()
+			allRefs = append(allRefs, refs...)
+			allRefsMu.Unlock()
```

Comment:
While this works, is there a reason you didn't choose to go with using a channel sending `refs` across that, and collecting results at the end? Appending to a shared memory slice protected by a mutex seems quite unidiomatic.